### PR TITLE
467 feature request add dsts support

### PR DESCRIPTION
**Pull request by handsomejack-42**

### CODE REVIEW

**Comment by handsomejack-42** [comment]
it's still in draft, but I would like some feedback @chlowell, if I'm taking this to the right direction. there's more than enough code already, I think I covered all the cases mentioned in the linked issue.

File: apps/confidential/confidential_test.go [hunk 1]
```diff
@@ -65,19 +66,19 @@ func TestCertFromPEM(t *testing.T) {
...
 	localhost         = "http://localhost"
 	refresh           = "fake_refresh"
 	token             = "fake_token"
 )
 
 var tokenScope = []string{"the_scope"}
 
-func fakeClient(tk accesstokens.TokenResponse, credential Credential, options ...Option) (Client, error) {
-	client, err := New(fakeAuthority, fakeClientID, credential, options...)
+func fakeClient(tk accesstokens.TokenResponse, credential Credential, fAuthority string, options ...Option) (Client, error) {
```
Comment (bgavrilMS) [inline apps/confidential/confidential_test.go:80]: Pls undo this name shortnening as it is harder to read the code.

File: apps/internal/oauth/ops/authority/authority_test.go [hunk 2]
```diff
@@ -320,22 +321,31 @@ func TestCreateAuthorityInfoFromAuthorityUri(t *testing.T) {
...
+
+		"override common to tenant":           {authority: host + "common", tenant: uuid1, expectedAuthority: host + uuid1},
+		"override organizations to tenant":    {authority: host + "organizations", tenant: uuid1, expectedAuthority: host + uuid1},
+		"override tenant to tenant2":          {authority: host + uuid1, tenant: uuid2, expectedAuthority: host + uuid2},
+		"override tenant to tenant2 for dSTS": {authority: host + "dstsv2/" + uuid1, tenant: uuid2, expectedAuthority: host + "dstsv2/" + uuid2},
+
+		"tenant can't be common for AAD":        {authority: host + uuid1, tenant: "common", expectError: true},
+		"tenant can't be consumers for AAD":     {authority: host + uuid1, tenant: "consumers", expectError: true},
+		"tenant can't be organizations for AAD": {authority: host + uuid1, tenant: "organizations", expectError: true},
+		"can't override tenant for ADFS ever":   {authority: host + "adfs", tenant: uuid1, expectError: true},
```
Comment (bgavrilMS) [inline apps/internal/oauth/ops/authority/authority_test.go:343]: Please add a test for ADFS where tenant is null / empty and where we expect to get back the original authority.

**Review by bgavrilMS — APPROVED** [review]

File: apps/internal/oauth/ops/authority/authority_test.go (hunk 2 above)
Comment (handsomejack-42) [reply apps/internal/oauth/ops/authority/authority_test.go:344]: added

File: apps/confidential/confidential_test.go (hunk 1 above)
Comment (handsomejack-42) [reply apps/confidential/confidential_test.go]: expanded

File: apps/internal/oauth/ops/authority/authority.go [hunk 3]
```diff
@@ -23,7 +23,8 @@ import (
 
 const (
 	authorizationEndpoint             = "https://%v/%v/oauth2/v2.0/authorize"
-	instanceDiscoveryEndpoint         = "https://%v/common/discovery/instance"
+	aadInstanceDiscoveryEndpoint      = "https://%v/common/discovery/instance"
+	dstsInstanceDiscoveryEndpoint     = "https://%v/dstsv2/common/discovery/instance"
```
Comment (victorm-hernandez) [inline apps/internal/oauth/ops/authority/authority.go:27]: >https://%v/dstsv2/common/discovery/instance [](http://example.com/codeflow?start=38&length=43)

It seems consuming this endpoint requires changes on the dSTS side. I believe this is the first time a client takes dependency on this endpoint with dSTS. 

If you see the output of the endpoint using a test dSTS instance you will notice that the metadata returned is incorrect and it points mostly to AAD.

Connect to the VPN before loading this on your browser:

[[https://co2agg04-dsts.dsts.core.azure-test.net/dstsv2/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2Fcommon%2Foauth2%2Fv2.0%2Fauthorize](https://co2agg04-dsts.dsts.core.azure-test.net/dstsv2/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2Fcommon%2Foauth2%2Fv2.0%2Fauthorize)](https://co2agg04-dsts.dsts.core.azure-test.net/dstsv2/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2Fcommon%2Foauth2%2Fv2.0%2Fauthorize)

This is the response

```json
{
    "tenant_discovery_endpoint": "[https://login.microsoftonline.com/dstsv2/common/v2.0/.well-known/openid-configuration](https://login.microsoftonline.com/dstsv2/common/v2.0/.well-known/openid-configuration)",
    "api-version": "1.1",
    "metadata": [
        {
            "preferred_network": "login.microsoftonline.com",
            "preferred_cache": "login.windows.net",
            "aliases": [
                "login.microsoftonline.com",
                "login.windows.net",
                "login.microsoft.com",
                "sts.windows.net"
            ]
        },
        {
            "preferred_network": "login.partner.microsoftonline.cn",
            "preferred_cache": "login.partner.microsoftonline.cn",
            "aliases": [
                "login.partner.microsoftonline.cn",
                "login.chinacloudapi.cn"
            ]
        },
        {
            "preferred_network": "login.microsoftonline.de",
            "preferred_cache": "login.microsoftonline.de",
            "aliases": [
                "login.microsoftonline.de"
            ]
        },
        {
            "preferred_network": "login.microsoftonline.us",
            "preferred_cache": "login.microsoftonline.us",
            "aliases": [
                "login.microsoftonline.us",
                "login.usgovcloudapi.net"
            ]
        },
        {
            "preferred_network": "login-us.microsoftonline.com",
            "preferred_cache": "login-us.microsoftonline.com",
            "aliases": [
                "login-us.microsoftonline.com"
            ]
        },
        {
            "preferred_network": "co2agg04-passive-dsts.dsts.core.azure-test.net/dstsv2",
            "preferred_cache": "co2agg04-passive-dsts.dsts.core.azure-test.net/dstsv2",
            "aliases": [
                "co2agg04-passive-dsts.dsts.core.azure-test.net"
            ]
        }
    ]
}
```

File: apps/internal/oauth/ops/authority/authority.go [hunk 4]
```diff
@@ -525,12 +533,24 @@ func (c Client) AADInstanceDiscovery(ctx context.Context, authorityInfo Info) (I
...
 		}
 
-		endpoint := fmt.Sprintf(instanceDiscoveryEndpoint, discoveryHost)
+		endpoint := fmt.Sprintf(aadInstanceDiscoveryEndpoint, discoveryHost)
 		err = c.Comm.JSONCall(ctx, endpoint, http.Header{}, qv, nil, &resp)
 	}
 	return resp, err
 }
 
+func (c Client) DSTSInstanceDiscovery(ctx context.Context, authorityInfo Info) (InstanceDiscoveryResponse, error) {
```
[the remark this answers is not in the document]
Comment (victorm-hernandez) [reply apps/internal/oauth/ops/authority/authority.go:542]: Correct, instance discovery is not supported.

File: apps/internal/oauth/ops/authority/authority.go [hunk 5]
```diff
@@ -344,44 +349,47 @@ type Info struct {
...
-	authorityType := AAD
-	if tenant == "adfs" {
+
+	var authorityType, tenant string
+	switch pathParts[1] {
+	case "adfs":
 		authorityType = ADFS
+	case "dstsv2":
+		if len(pathParts) != 3 {
+			return Info{}, errors.New(`dSTS authority must be an https URL such as "https://<authority>/dstsv2/<your tenant>"`)
```
Comment (victorm-hernandez) [inline apps/internal/oauth/ops/authority/authority.go:379]: ><your tenant> [](http://example.com/codeflow?start=102&length=13)

dSTS doesnt support tenants, the value for the tenant is hard coded everywhere and should always be the same. 

Lets change the error to reflect this. The tenant is

7a433bfc-2514-4697-b467-e0933190487f

File: apps/internal/oauth/resolvers.go [hunk 6]
```diff
@@ -119,9 +119,15 @@ func (m *authorityEndpoint) addCachedEndpoints(authorityInfo authority.Info, use
 	m.cache[authorityInfo.CanonicalAuthorityURI] = updatedCacheEntry
 }
 
-func (m *authorityEndpoint) openIDConfigurationEndpoint(ctx context.Context, authorityInfo authority.Info, userPrincipalName string) (string, error) {
-	if authorityInfo.Tenant == "adfs" {
+func (m *authorityEndpoint) openIDConfigurationEndpoint(ctx context.Context, authorityInfo authority.Info) (string, error) {
+	if authorityInfo.AuthorityType == authority.ADFS {
 		return fmt.Sprintf("https://%s/adfs/.well-known/openid-configuration", authorityInfo.Host), nil
+	} else if authorityInfo.AuthorityType == authority.DSTS {
+		resp, err := m.rest.Authority().DSTSInstanceDiscovery(ctx, authorityInfo)
```
Comment (victorm-hernandez) [inline apps/internal/oauth/resolvers.go:126]: >DSTSInstanceDiscovery [](http://example.com/codeflow?start=34&length=21)

Similar to ADFS, you can use a hard coded URL for dSTS. An example of this endpoint on dSTS

[https://test-instance1-dsts.dsts.core.azure-test.net/dstsv2/7a433bfc-2514-4697-b467-e0933190487f/.well-known/openid-configuration](https://test-instance1-dsts.dsts.core.azure-test.net/dstsv2/7a433bfc-2514-4697-b467-e0933190487f/.well-known/openid-configuration)

IMPORTANT! To connect to this endpoint you need to be on our VPN.

File: apps/internal/oauth/ops/authority/authority_test.go [hunk 7]
```diff
@@ -320,22 +321,32 @@ func TestCreateAuthorityInfoFromAuthorityUri(t *testing.T) {
...
 
 func TestAuthParamsWithTenant(t *testing.T) {
-	uuid1 := "00000000-0000-0000-0000-000000000000"
-	uuid2 := strings.ReplaceAll(uuid1, "0", "1")
+	uuid1 := uuid.New().String()
+	uuid2 := uuid.New().String()
 	host := "https://localhost/"
-	for _, test := range []struct {
+
+	tests := map[string]struct {
```
Comment (victorm-hernandez) [inline apps/internal/oauth/ops/authority/authority_test.go:328]: >tests [](http://example.com/codeflow?start=1&length=5)

Please make sure to run this against real dSTS at least once. 

The best way to do this is to create a sample project on the following [repository](https://msazure.visualstudio.com/One/_git/EngSys-Security-dSTS?path=/Samples/Current)

File: apps/confidential/confidential_test.go [hunk 8]
```diff
@@ -1328,3 +1333,72 @@ func TestWithAuthenticationScheme(t *testing.T) {
...
+			if err != nil {
+				t.Fatal(err)
+			}
+			client, err := fakeClient(accesstokens.TokenResponse{
+				AccessToken:   token,
+				ExpiresOn:     internalTime.DurationTime{T: time.Now().Add(1 * time.Hour)},
+				ExtExpiresOn:  internalTime.DurationTime{T: time.Now().Add(1 * time.Hour)},
+				GrantedScopes: accesstokens.Scopes{Slice: tokenScope},
+				TokenType:     "Bearer",
+			}, cred, "https://fake_authority/dstsv2/fake_tenant")
```
Comment (victorm-hernandez) [inline apps/confidential/confidential_test.go:1357]: >fake_tenant [](http://example.com/codeflow?start=43&length=11)

ditto: tenant is always fixed on dSTS

File: apps/confidential/confidential_test.go [hunk 9]
```diff
@@ -1328,3 +1333,72 @@ func TestWithAuthenticationScheme(t *testing.T) {
...
+
+			tk, err = client.AcquireTokenSilent(context.Background(), tokenScope)
+			if err != nil {
+				t.Errorf("got err == %s, want err == nil", err)
+			}
+			if tk.AccessToken != token {
+				t.Errorf("unexpected access token %s", tk.AccessToken)
+			}
+
+			// fail for another tenant
```
Comment (victorm-hernandez) [inline apps/confidential/confidential_test.go:1384]: >tenant [](http://example.com/codeflow?start=23&length=6)

Given that dSTS has a single tenant, there is no need for this test.

File: apps/internal/oauth/ops/authority/authority.go [hunk 10]
```diff
@@ -236,23 +238,26 @@ func NewAuthParams(clientID string, authorityInfo Info) AuthParams {
...
-	if p.AuthorityInfo.Tenant == "consumers" {
-		return p, errors.New(`client is configured to authenticate only personal Microsoft accounts, via the "consumers" endpoint`)
+		if p.AuthorityInfo.Tenant == "consumers" {
+			return p, errors.New(`client is configured to authenticate only personal Microsoft accounts, via the "consumers" endpoint`)
+		}
+		authority = "https://" + path.Join(p.AuthorityInfo.Host, ID)
+	case ADFS:
+		return p, errors.New("ADFS authority doesn't support tenants")
+	case DSTS:
+		authority = "https://" + path.Join(p.AuthorityInfo.Host, "dstsv2", ID)
```
Comment (victorm-hernandez) [inline apps/internal/oauth/ops/authority/authority.go:258]: >authority [](http://example.com/codeflow?start=2&length=9)

Since the tenant ID should always be the same for dSTS, the correct behavior is to throw if it is not the expected GUID

7a433bfc-2514-4697-b467-e0933190487f

otherwise build the authority URL.

**Review by victorm-hernandez — CHANGES_REQUESTED** [review]
:clock1:

File: apps/internal/oauth/ops/authority/authority.go (hunk 3 above)
Comment (handsomejack-42) [reply apps/internal/oauth/ops/authority/authority.go:27]: removed

File: apps/internal/oauth/resolvers.go (hunk 6 above)
Comment (handsomejack-42) [reply apps/internal/oauth/resolvers.go:126]: hard-coded the url

File: apps/confidential/confidential_test.go (hunk 8 above)
Comment (handsomejack-42) [reply apps/confidential/confidential_test.go:1357]: changed

File: apps/internal/oauth/ops/authority/authority.go (hunk 10 above)
Comment (handsomejack-42) [reply apps/internal/oauth/ops/authority/authority.go:258]: changed to always return an error if it hits this `case`, there are only three possible cases on `WithTenant` for dSTS flow that we care about

1. `ID == ""` -> returns identical `AuthParams`
2. `ID == dSTS tenant` -> returns identical AuthParams
3. neither of the above -> should return error

File: apps/internal/oauth/ops/authority/authority_test.go (hunk 7 above)
Comment (handsomejack-42) [reply apps/internal/oauth/ops/authority/authority_test.go:328]: this test isn't actually firing any http requests, can you elaborate on what you mean by 'against real dSTS at least once'? I am not aware of any tests which I changed in the context of this PR are actually accessing anything e2e

Comment (handsomejack-42) [reply apps/internal/oauth/ops/authority/authority_test.go:328]: I mean if you want me to manually validate that we can get a token from dSTS using this flow I can do that (I already did that a long time ago, but the set-up we used was deprecated by the recent dSTS crack-down on MS-corp based identities)

I guess that is why you are suggesting to 'build my own test dSTS' in C# and run it locally to test against? because at this point I can't see any easy way of how to call dSTS from SAW (I can hardly compile go there), so the only thing that comes to mind is some ev2 shell extension or something

Comment (bgavrilMS) [reply apps/internal/oauth/ops/authority/authority_test.go:328]: I recommend moving this discussion outside of GitHub. Happy to cut a preview release for testing purposes. Agree that we should have some sort of manual validation, at least of some scenarios (client_credentials flow probably. I think user flows are a bit broken right now anyway)

Comment (handsomejack-42) [reply apps/internal/oauth/ops/authority/authority_test.go:328]: done

**Comment by handsomejack-42** [comment]
ready to be merged, ran against real dSTS instance as instructed, works

@bgavrilMS @victorm-hernandez

File: apps/confidential/confidential_test.go [hunk 11]
```diff
@@ -21,15 +21,16 @@ import (
 	"testing"
 	"time"
 
+	"github.com/golang-jwt/jwt/v5"
```
Comment (bgavrilMS) [inline apps/confidential/confidential_test.go:24]: Can you please remove these 2 imports? I do not even see them being used.
